core/commands/filestore.go

```diff
@@ -0,0 +1,946 @@
+	fsutil "github.com/ipfs/go-ipfs/filestore/util"
+	"github.com/ipfs/go-ipfs/repo/fsrepo"
+)
+
+var FileStoreCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Interact with filestore objects",
+	},
+	Subcommands: map[string]*cmds.Command{
+		"add":      addFileStore,
+		"ls":       lsFileStore,
+		"ls-files": lsFiles,
+		"verify":   verifyFileStore,
+		"rm":       rmFilestoreObjs,
+		"clean":    cleanFileStore,
+		"fix-pins": repairPins,
+		"unpinned": fsUnpinned,
+		"rm-dups":  rmDups,
+		"upgrade":  fsUpgrade,
+		"mv":       moveIntoFilestore,
+	},
+}
+
+var addFileStore = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Add files to the filestore.",
+		ShortDescription: `
+Add contents of <path> to the filestore.  Most of the options are the
+same as for "ipfs add".
+`},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("path", true, true, "The path to a file to be added."),
+	},
+	Options: addFileStoreOpts(),
+	PreRun: func(req cmds.Request) error {
+		serverSide, _, _ := req.Option("server-side").Bool()
+		logical, _, _ := req.Option("logical").Bool()
+		physical, _, _ := req.Option("physical").Bool()
+		if logical && physical {
+			return errors.New("Both --logical and --physical can not be specified.")
+		}
+		cwd := ""
+		var err error
+		if logical {
+			cwd, err = filestore.EnvWd()
+		}
+		if physical {
+			cwd, err = filestore.SystemWd()
+		}
+		if err != nil {
+			return err
+		}
+		if cwd != "" {
+			paths := req.Arguments()
+			for i, path := range paths {
+				abspath, err := filestore.AbsPath(cwd, path)
+				if err != nil {
+					return err
+				}
+				paths[i] = abspath
+			}
+			req.SetArguments(paths)
+		}
+		if !serverSide {
+			err := getFiles(req)
+			if err != nil {
+				return err
+			}
+		}
+		return AddCmd.PreRun(req)
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		node, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		config, _ := req.InvocContext().GetConfig()
+		serverSide, _, _ := req.Option("server-side").Bool()
+		if serverSide && !config.Filestore.APIServerSidePaths {
+			res.SetError(errors.New("Server Side Adds not enabled."), cmds.ErrNormal)
+			return
+		}
+		if serverSide {
+			err := getFiles(req)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+		} else if node.OnlineMode() {
+			if !req.Files().IsDirectory() {
+				res.SetError(errors.New("expected directory object"), cmds.ErrNormal)
+				return
+			}
+			req.SetFiles(&fixPath{req.Arguments(), req.Files()})
+		}
+		req.Values()["no-copy"] = true
+		AddCmd.Run(req, res)
+	},
+	PostRun: AddCmd.PostRun,
+	Type:    AddCmd.Type,
+}
+
+func addFileStoreOpts() []cmds.Option {
+	var opts []cmds.Option
+	opts = append(opts, AddCmd.Options...)
+	opts = append(opts,
+		cmds.BoolOption("server-side", "S", "Read file on server."),
+		cmds.BoolOption("l", "logical", "Create absolute path using the PWD from environment."),
```

Comment:
`PWD` is a environmental variable from the shell, spelling out would be incorrect.